File: arduino-mega/2inch13_epaper/epd_2inch13.cpp
#include <SPI.h>
#include <string.h>
#include <stdlib.h>		

#include "epd_2inch13.h"

unsigned char Border_Flag = 0;
//LUT_Flag is the call waveform flag
unsigned char LUT_Flag=0;


//Configure the serial port to use the standard printf function
//start
int serial_putc( char c, struct __file * )
{
  Serial.write( c );
  return c;
}
void printf_begin(void)
{
  fdevopen( &serial_putc, 0 );
}
//end

/***************************************************************
function:Hardware initialization
***************************************************************/
void Epd_gpio_init(void)
{
    //gpio init
    pinMode(PIN_CS, OUTPUT);
    pinMode(PIN_RST, OUTPUT);
    pinMode(PIN_DC, OUTPUT);
    pinMode(PIN_TEST, OUTPUT);
    pinMode(PIN_BUSY, INPUT_PULLUP);
    //Serial init
    Serial.begin(115200);
    printf_begin();
    printf("Init gpio for arduino\r\n"); 
    //spi init
#ifdef SPI_USE_HARDWARE
    SPI.begin();
    SPI.setDataMode(SPI_MODE3);
    SPI.setBitOrder(MSBFIRST);
    SPI.setClockDivider(SPI_CLOCK_DIV4);
    SPI.begin();
    printf("use hardware spi\r\n");
#else
    pinMode(PIN_MOSI, OUTPUT);
    pinMode(PIN_CLK, OUTPUT); 
    printf("use gpio mode spi\r\n");
#endif
}

/***************************************************************
function:Hardware reset
***************************************************************/
void Epd_reset(void)
{
  digitalWrite(PIN_CS, LOW);
	digitalWrite(PIN_RST, HIGH);
	delay(10);   //At least 10ms delay 
	digitalWrite(PIN_RST, LOW);
	delay(100);  //At least 10ms delay 
	digitalWrite(PIN_RST, HIGH);
	delay(100);  //At least 10ms delay 
}

#ifndef SPI_USE_HARDWARE
void SPI_transfer(unsigned char value)                                    
{                                                           
  unsigned char i;
  for(i=0; i<8; i++)   
  {
      digitalWrite(PIN_CLK, LOW);
      if(value & 0x80)
        digitalWrite(PIN_MOSI, HIGH);
      else
        digitalWrite(PIN_MOSI, LOW);  
      value = (value << 1);     
      digitalWrite(PIN_CLK, HIGH);      
  }
}
#endif
/****************************************************************
function:Write data and commands
*****************************************************************/
static void Epd_write_cmd(unsigned char data)	 
{	
    uint8_t res=0;
  	digitalWrite(PIN_CS, LOW);
  	digitalWrite(PIN_DC, LOW); // command write
#ifdef SPI_USE_HARDWARE
  	res = SPI.transfer(data);
#else 
    SPI_transfer(data);
#endif
    digitalWrite(PIN_CS, HIGH);
    digitalWrite(PIN_DC, HIGH); 
}

void Epd_write_data(unsigned char data) 
{	
    uint8_t res=0;
    digitalWrite(PIN_CS, LOW);
    digitalWrite(PIN_DC, HIGH); // data write
#ifdef SPI_USE_HARDWARE
    res = SPI.transfer(data);
#else
    SPI_transfer(data);
#endif
    digitalWrite(PIN_CS, HIGH);
    digitalWrite(PIN_DC, HIGH); 
}    


/******************************************************************************
function:	Common register initialization
******************************************************************************/
void Epd_init(void)
{
    LUT_Flag = 0;

    Epd_write_cmd(0x00);      
    Epd_write_data(0xFB);//Can change the scanning direction,FB->rotation  F3->left/right F7->up/down
    Epd_write_data(0x89); 
    
    Epd_write_cmd(0x01); 
    Epd_write_data(0x03);    
    Epd_write_data(0x00);    
    Epd_write_data(0x3F);    
    Epd_write_data(0x3F);    
    Epd_write_data(0x03);    

    Epd_write_cmd(0x03);       
    Epd_write_data(0x00);    

    Epd_write_cmd(0x06);       
    Epd_write_data(0x27);    
    Epd_write_data(0x27);    
    Epd_write_data(0x2F);    

    Epd_write_cmd(0x30);       
    Epd_write_data(0x09);   

    Epd_write_cmd(0x60);       
    Epd_write_data(0x22);   

    Epd_write_cmd(0x82);    // vcom
    Epd_write_data(0x00);   // 

    Epd_write_cmd(0xE3);        
    Epd_write_data(0x00);    

    Epd_write_cmd(0x41);        
    Epd_write_data(0x00);   
    
    Epd_write_cmd(0x61);    
    Epd_write_data(0x80);     
    Epd_write_data(0x00);   
    Epd_write_data(0xFA);  
  
    Epd_write_cmd(0x65);        
    Epd_write_data(0x00);    
    Epd_write_data(0x00);    
    Epd_write_data(0x00);    

    Epd_write_cmd(0X50);      
    Epd_write_data(0xB7); 
}

//***************Epd_chkstatus***********************//  
//GPIO reads BUSY pin,
//Pin level=L indicates that the screen driver chip is busy and cannot send data and command
//Pin level=H indicates that the screen driver chip is idle and can send data and command
//***************************************************//
void Epd_chkstatus(void)
{
   while(digitalRead(PIN_BUSY)==0); 
}

//****************EPD_refresh(void)**********************//  
//Screen refresh instruction
//******************************************************//
void Epd_refresh(void)
{
    Border_Flag=1;  
    Epd_write_cmd(0x17);      
    Epd_write_data(0xA5);
    Epd_chkstatus();
} 

//***************EPD_sleep(void)************************//  
//enter sleep
//After exiting sleep, it is necessary to reinitialize the settings. 
//The first screen must be fully refresh by GC, and then the screen can be refresh by DU
//******************************************************//
void Epd_sleep(void)
{
    Epd_write_cmd(0X07); //deep sleep
    Epd_write_data(0xA5);
}

//***************lut_GC(void)************************//  
//Call GC waveform:Global Refresh
//***************************************************//
void lut_GC(void)
{
   unsigned int count;  
  
   Epd_write_cmd(0x20);             //vcom
   for(count=0;count<56;count++)
      {Epd_write_data(lut_R20_GC[count]);}
    
   Epd_write_cmd(0x21);             //red not use
   for(count=0;count<56;count++)
      {Epd_write_data(lut_R21_GC[count]);}

   Epd_write_cmd(0x24);             //bb b
   for(count=0;count<56;count++)
      {Epd_write_data(lut_R24_GC[count]);}     
      
   if (LUT_Flag==0) 
   { 
        Epd_write_cmd(0x22);             //bw r
        for(count=0;count<56;count++)
          {Epd_write_data(lut_R22_GC[count]);}
    
        Epd_write_cmd(0x23);             //wb w
        for(count=0;count<56;count++)
          {Epd_write_data(lut_R23_GC[count]);}
        LUT_Flag=1;
   }
   else
   {
      Epd_write_cmd(0x22);             //bw r
      for(count=0;count<56;count++)
        {Epd_write_data(lut_R23_GC[count]);}
  
      Epd_write_cmd(0x23);             //wb w
      for(count=0;count<56;count++)
        {Epd_write_data(lut_R22_GC[count]);}
      LUT_Flag=0;
   }
                 
}

//***************lut_DU(void)************************//  
//Call DU waveform:Partial refresh
//***************************************************//
void lut_DU(void)
{
    unsigned int count;  
  
    Epd_write_cmd(0x20);             //vcom
    for(count=0;count<56;count++)
      {Epd_write_data(lut_R20_DU[count]);}
    
    Epd_write_cmd(0x21);             //red not use
    for(count=0;count<56;count++)
      {Epd_write_data(lut_R21_DU[count]);}

    Epd_write_cmd(0x24);             //bb b
    for(count=0;count<56;count++)
      {Epd_write_data(lut_R24_DU[count]);}     
      
   if (LUT_Flag==0) 
   {
      Epd_write_cmd(0x22);             //bw r
      for(count=0;count<56;count++)
        {Epd_write_data(lut_R22_DU[count]);}
  
      Epd_write_cmd(0x23);             //wb w
      for(count=0;count<56;count++)
        {Epd_write_data(lut_R23_DU[count]);}
      LUT_Flag=1;
   }
   else
   {
      Epd_write_cmd(0x22);             //bw r
      for(count=0;count<56;count++)
        {Epd_write_data(lut_R23_DU[count]);}
  
      Epd_write_cmd(0x23);             //wb w
      for(count=0;count<56;count++)
        {Epd_write_data(lut_R22_DU[count]);}
      LUT_Flag=0;
   }
                 
}

//***************PIC_display_DU**********************//  
//Call DU waveform to refresh screen
//***************************************************//
void PIC_display_DU(const unsigned char* picData)
{
    unsigned int i;
  
    Epd_write_cmd(0X50);   // 
    Epd_write_data(0xD7);  // Border 
   
    Epd_write_cmd(0x13);   //Transfer new data
    for(i=0;i<(EPD_MEM_W*EPD_MEM_H/8);i++)       
    {
          Epd_write_data(*picData);
          picData++;
    } 
        
    lut_DU();           
    Epd_refresh();            
}

//***************PIC_display_GU**********************//  
//Call GC waveform to refresh screen
//***************************************************//
void PIC_display_GC(const unsigned char* picData)
{
    unsigned int i;
    
    //Border_Flag refers to the border mark. Only the first screen after initialization will be painted with Border, 
    //and other times will not be painted
    if (Border_Flag==1) 
    {
         Epd_write_cmd(0X50);      // 
         Epd_write_data(0xD7);     //  not Border
    }  
    //***********************************************************************//
    Epd_write_cmd(0x13);         
    for(i=0;i<(EPD_MEM_W*EPD_MEM_H/8);i++)       
    {
         Epd_write_data(*picData);
         picData++;
    } 

    lut_GC();         
    Epd_refresh();          
}

File: raspberry_pi/python/epd_2inch13.py
# Seengreat 2.13 Inch E-Paper Display demo
# Author(s):Andy Li from Seengreat

import os
import sys
import spidev
import time
import wiringpi as wpi
import numpy as np
"""the following pin definiting use wiringpi"""
PIN_CS     = 10
PIN_DC     = 6
PIN_BUSY   = 5
PIN_RST    = 0
# DU waveform

#GC refresh :800ms
lut_R20_GC = [ 0x01,0x00,0x14,0x14,0x01,0x01,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
lut_R21_GC = [ 0x01,0x60,0x14,0x14,0x01,0x01,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
lut_R22_GC = [ 0x01,0x60,0x14,0x14,0x01,0x01,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
lut_R23_GC = [ 0x01,0x90,0x14,0x14,0x01,0x01,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
lut_R24_GC = [ 0x01,0x90,0x14,0x14,0x01,0x01,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]

#DU refreah: 400ms
lut_R20_DU = [ 0x01,0x00,0x14,0x01,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
lut_R21_DU = [ 0x01,0x20,0x14,0x01,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
lut_R22_DU = [ 0x01,0x80,0x14,0x01,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
lut_R23_DU = [ 0x01,0x40,0x14,0x01,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]
lut_R24_DU = [ 0x01,0x00,0x14,0x01,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
               0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00]

class EPD_2Inch13():
    def __init__(self):
        # gpio init
        wpi.wiringPiSetup()
        wpi.pinMode(PIN_DC, wpi.OUTPUT)  # D/C pin
        wpi.pinMode(PIN_RST, wpi.OUTPUT)  # reset pin
        wpi.pinMode(PIN_CS, wpi.OUTPUT)  # cs pin
        wpi.pinMode(PIN_BUSY, wpi.INPUT)  # BUSY pin
        # spi init
        self.bus = 0
        self.dev = 0
        self.spi_speed = 8000000
        self.spi = spidev.SpiDev()
        self.spi.open(self.bus, self.dev)
        self.spi.max_speed_hz = self.spi_speed
        self.spi.mode = 0b00

        self.mem_w = 250
        self.mem_h = 128
        
        self.dis_w = 250
        self.dis_h = 122
        
        self.Border_flag = 0
        self.LUT_lfag = 0
        
    def write_cmd(self, cmd):
        """write command"""
        wpi.digitalWrite(PIN_DC,wpi.LOW)
        self.spi.writebytes([cmd])
        
    def write_data(self, value):
        """write data"""
        wpi.digitalWrite(PIN_DC,wpi.HIGH)
        self.spi.writebytes([value])
        
    def chkstatus(self):
        while wpi.digitalRead(PIN_BUSY)==wpi.LOW:
            pass
        
    def reset(self):
        """reset the epd"""
        wpi.digitalWrite(PIN_RST,wpi.HIGH)
        time.sleep(0.01)
        wpi.digitalWrite(PIN_RST,wpi.LOW)
        time.sleep(0.1)
        wpi.digitalWrite(PIN_RST,wpi.HIGH)
        time.sleep(0.1)
        
    def epd_init(self):
        """epd init..."""
        self.LUT_flag = 0
        
        self.write_cmd(0x00)
        self.write_data(0xFB)
        self.write_data(0x89)
        
        self.write_cmd(0x01)      
        self.write_data(0x03)
        self.write_data(0x00)
        self.write_data(0x3F)
        self.write_data(0x3F)
        self.write_data(0x03)        

        self.write_cmd(0x03)
        self.write_data(0x00)
        
        self.write_cmd(0x06)
        self.write_data(0x27)
        self.write_data(0x27)
        self.write_data(0x2F)
        
        self.write_cmd(0x30) 
        self.write_data(0x09)
        
        self.write_cmd(0x60) 
        self.write_data(0x22)
        
        self.write_cmd(0x82) 
        self.write_data(0x00)
        
        self.write_cmd(0xE3) 
        self.write_data(0x00)
        
        self.write_cmd(0x41) 
        self.write_data(0x00)
        
        self.write_cmd(0x61) 
        self.write_data(0x80)
        self.write_data(0x00)
        self.write_data(0xFA)
        
        self.write_cmd(0x65) 
        self.write_data(0x00)
        self.write_data(0x00)
        self.write_data(0x00)
        
        self.write_cmd(0x50) 
        self.write_data(0xB7)

    def refresh(self):
        self.Border_flag = 1
        self.write_cmd(0x17) 
        self.write_data(0xA5)
        self.chkstatus()
        
    def sleep(self):
        self.write_cmd(0x07)
        self.write_data(0xA5)
        
    #Call GC waveform:Global Refresh
    def lut_GC(self):
        self.write_cmd(0x20) # vcom
        for count in range(0,56):
            self.write_data(lut_R20_GC[count])
        self.write_cmd(0x21)# red not use
        for count in range(0,56):
            self.write_data(lut_R21_GC[count])
        self.write_cmd(0x24) # bb b
        for count in range(0,56):
            self.write_data(lut_R24_GC[count])

        if self.LUT_flag==0:
            self.write_cmd(0x22) #bw r
            for count in range(0,56):
                self.write_data(lut_R22_GC[count])
            self.write_cmd(0x23) #wb w
            for count in range(0,56):
                self.write_data(lut_R23_GC[count])
            self.LUT_flag=1
        else:
            self.write_cmd(0x22) # bw r
            for count in range(0,56):
                self.write_data(lut_R23_GC[count])
            self.write_cmd(0x23) #wb w
            for count in range(0,56):
                self.write_data(lut_R22_GC[count])
            self.LUT_flag=0
            
    #Call DU waveform:Partial Refresh
    def lut_DU(self):
        self.write_cmd(0x20) # vcom
        for count in range(0,56):
            self.write_data(lut_R20_DU[count])
        self.write_cmd(0x21)# red not use
        for count in range(0,56):
            self.write_data(lut_R21_DU[count])
        self.write_cmd(0x24) # bb b
        for count in range(0,56):
            self.write_data(lut_R24_DU[count])

        if self.LUT_flag==0:
            self.write_cmd(0x22) #bw r
            for count in range(0,56):
                self.write_data(lut_R22_DU[count])
            self.write_cmd(0x23) #wb w
            for count in range(0,56):
                self.write_data(lut_R23_DU[count])
            self.LUT_flag=1
        else:
            self.write_cmd(0x22) # bw r
            for count in range(0,56):
                self.write_data(lut_R23_DU[count])
            self.write_cmd(0x23) #wb w
            for count in range(0,56):
                self.write_data(lut_R22_DU[count])
            self.LUT_flag=0
            
    #Call DU waveform to refresh screen
    def display_DU(self,picData):
        self.write_cmd(0X50)
        self.write_data(0xD7)# Border
        self.write_cmd(0x13)#Transfer new data
        for i in range(0,int(self.mem_w*self.mem_h/8)):
            self.write_data(picData[i])
        self.lut_DU()
        self.refresh()

    #Call GC waveform to refresh screen
    def display_GC(self, picData):
        '''Border_Flag refers to the border mark. Only the first screen after initialization
        will be painted with Border,and other times will not be painted'''
        if self.Border_flag==1:
            self.write_cmd(0X50)
            self.write_data(0xD7) # not Border
        self.write_cmd(0x13)
        for i in range(0,int(self.mem_w*self.mem_h/8)):
            self.write_data(picData[i])
        self.lut_GC()
        self.refresh()


    